### ISSUE TITLE
interoperability with pyvisa

### ISSUE THREAD

**Issue by arsenovic** [p3trus/slave#140]
i have some classes written for instruments using pyvisa,

https://github.com/scikit-rf/scikit-rf/tree/master/skrf/vi

 but would like to be able to use them with slave. however i notice that  when using the gpib device it doesnt handle incorrectly written queries, like ask('*INsD?') (it seems to lockup the device unit i gpib_config it again). 

also the pyvisa gpib class has a handy method for ask_for_values (http://pyvisa.readthedocs.org/en/latest/instruments.html#Instrument.ask_for_values), which would be needed for seemless interoperability

any thoughts on this?

**Comment by arsenovic** [comment]
i was just about to re-create this issue. i had forgotten i started it in april!

anyway, i see your Command class requires the returned values to be Type'd. so it seems to me that the correct point for this feature would be to make a Type for array, or list ?

**Comment by p3trus** [comment]
If you know the exact number of return types you can use more than one type. E.g. let's say the command takes 2 floats and returns 2 floats it would look like this:

    from slave.driver import Driver, Command
    from slave.types import Float
    import itertools

    class MyDriver(Driver):
        def __init__(self, transport):
            super(MyDriver, self).__init__(transport)
            cmd = Command('QRY?', 'WRITE', [Float, Float])

If you have a variable number of types you can use an infinite iterator of types e.g.

    from slave.driver import Driver, Command
    from slave.types import Float
    import itertools

    class MyDriver(Driver):
        def __init__(self, transport):
            super(MyDriver, self).__init__(transport)
            variable_cmd = Command('QRY?', 'WRITE', itertools.repeat(Float()))

**Comment by arsenovic** [comment]
ok thanks, ill give that a try. 

seems like  creating a list/array Type would be  useful for instruments  like spectrum analyzers, vector network analyzers, or anything that returns trace-like data. this seems to me more natural than an iterated float.

**Comment by p3trus** [comment]
You're right it looks ugly but the issue didn't came up very often in my use cases. I've got a few questions:
 * Do you have mixed return types in your usecases? E.g. int, float pairs?
 * Does the number of returned values vary?

I guess something like `Command('QRY?', 'WRITE', Sequence(Integer, Float))` would be nice. I will think about it. Shouldn't be too hard to implement.

**Comment by arsenovic** [comment]
usually,  the returned value is a variable-length, delimited list of a single type. 
`3.4,1.1, 1231.1,....`

 i added the following function to python-ivi, which is compatible with the previous pyvisa command. 


```
from ivi import Driver 
def ask_for_values(self,msg, delim=',', converter=float, array = True):
    s = self._ask(msg)
    s_split = s.split(delim)
    out = map(converter, s_split)
    if array:
        out = npy.array(out)
    return out

Driver.ask_for_values = ask_for_values
```


however, to use this correctly with slave it belongs as a Type. in my case, the list is all floats but represents a flattened array of complex numbers. (`r,i,r,i,r,i...`)

however, i thought this level of processing  is too specialized to put into slave.

**Comment by p3trus** [comment]
A simple first implementation would be something like this

    from slave.driver import _to_instance
    import itertools

    class Stream(object):
        def __init__(self, *types):
            self.types = map(_to_instance, types)
        
        def __iter__(self):
            return itertools.cycle(self.types)

You could use it like this
    
    from slave.driver import Driver, Command
    from slave.types import Float, String, Integer
    import itertools

    class MyDriver(Driver):
        def __init__(self, transport):
            super(MyDriver, self).__init__(transport)
            # A command with a variing number of types alternating between floats and ints.
            cmd = Command('QUERY?', 'WRITE', Stream(Float, Integer))

I'm probably adding this to the repo but don't have the time today to write unittest etc.

**Comment by arsenovic** [comment]
sweet thanks!

**Comment by p3trus** [comment]
For testing purposes `MockTransport` might be helpful. Using it with the above code

    from slave.test.test_protocol import MockTransport

    transport = MockTransport(responses=['12.5,12,13.5,13\n'])
    driver = MyDriver(transport)
    # Test queriing
    print "Let's query it:", driver.cmd
    print "And the message was:", transport.messages[-1]
    # Test writing
    driver.cmd = 12.2, 134, 143.3
    print "And the message was:", transport.messages[-1]
